lib/FrontendTool/FrontendTool.cpp

```diff
@@ -262,6 +264,313 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+  os << "error: invariant violation: ";
+  f(os);
+  llvm::report_fatal_error(os.str());
+#endif
+}
+
+// [Note: Trace-Clang-submodule-complexity]
+//
+// A Clang module may have zero or more submodules. In practice, when traversing
+// the imports of a module, we observe that different submodules of the same
+// top-level module (almost) freely import each other. Despite this, we still
+// need to conceptually traverse the tree formed by the submodule relationship
+// (with the top-level module being the root).
+//
+// This needs to be taken care of in two ways:
+// 1. We need to make sure we only go towards the leaves. It's okay if we "jump"
+//    branches, so long as we don't try to visit an ancestor when one of its
+//    descendants is still on the traversal stack, so that we don't end up with
+//    arbitrarily complex intra-module cycles.
+//    See also: [Note: Intra-module-leafwards-traversal].
+// 2. When adding entries to the ABI export map, we need to avoid marking
+//    dependencies within the same top-level module. This step is needed in
+//    addition to step 1 to avoid creating cycles like
+//    Overlay -> Underlying -> Submodule -> Overlay.
+
+void ABIDependencyEvaluator::addToABIExportMap(
+    ModuleDecl *module, ModuleDecl *reexport) {
+  if (module == reexport) {
+    crashOnInvariantViolation([&](llvm::raw_string_ostream &os){
+      os << "expected module "; printModule(reexport, os);
+      os << "  to not re-export itself\n";
+    });
+    return;
+  }
+  if (reexport->isNonSwiftModule()
+      && module->isNonSwiftModule()
+      && module->getTopLevelModule() == reexport->getTopLevelModule()) {
+    // Dependencies within the same top-level Clang module are not useful.
+    // See also: [Note: Trace-Clang-submodule-complexity].
+    return;
+  }
+
+  // We only care about dependencies across top-level modules and we want to
+  // avoid exploding abiExportMap with submodules. So we only insert entries
+  // after calling getTopLevelModule().
+
+  if (::isClangOverlayOf(module, reexport)) {
+    // For overlays, we need to have a dependency on the underlying module.
+    // Otherwise, we might accidentally create a Swift -> Swift cycle.
+    abiExportMap[module].insert(
+      reexport->getTopLevelModule(/*preferOverlay*/false));
+    return;
+  }
+  abiExportMap[module].insert(
+      reexport->getTopLevelModule(/*preferOverlay*/true));
+}
+
+void ABIDependencyEvaluator::reexposeImportedABI(
+    ModuleDecl *module, ModuleDecl *importedModule) {
+  if (module == importedModule) {
+    crashOnInvariantViolation([&](llvm::raw_string_ostream &os) {
+      os << "module "; printModule(module, os); os << " imports itself!\n";
+    });
+    return;
+  }
+  computeABIDependenciesForModule(importedModule);
+  addToABIExportMap(module, importedModule);
+  auto &reexportsRef = abiExportMap[importedModule];
+  llvm::DenseSet<ModuleDecl *> reexports(reexportsRef);
+  for (auto reexportedModule: reexports)
```

Comment:
This might seem odd: why are we copying the set instead of iterating over it directly in the loop with `auto reexportedModule: abiExportMap[importedModule]`? After all, we already checked that `module != importedModule`, so this set should not be modified in the calls to `addToABIExportMap`.

I'm afraid I don't really understand why this is needed.

However, it does seem to fix the difficult-to-reproduce iterator invalidation I was seeing earlier for `DenseMap` when building llbuildSwift and SwiftPM. (More specifically: I hit an assertion error in `operator++` with `isHandleInSync()`.)

(This pattern is repeated in one another place.)

Reply:
Huh that's interesting, it seems like `DenseMap` (the underlying storage of `DenseSet`) is invalidating its iterator when it gets moved during a re-allocation of the `DenseMap` it's stored in. Its move constructor calls the `swap` method which calls `incrementEpoch`, signalling an iterator invalidation.

I'm not really sure why it does this though, it seems like the iterator should be able to survive a move, perhaps `DenseMap` doesn't want to formally guarantee this though? Unfortunately there doesn't appear to be much documentation on this :/

Reply:
> is invalidating its iterator when it gets moved during a re-allocation of the DenseMap it's stored in. Its move constructor calls the swap method which calls incrementEpoch, signalling an iterator invalidation.

But why is anything getting moved at all (i.e. I'm not sure how you arrived at the conclusion that there's a move happening here)? I would've thought the semantics of `auto reexportedModule: abiExportMap[importedModule]` would be that we iterate over the existing DenseSet entry without moving anything.

Reply:
I'm looking at [cppreference](https://en.cppreference.com/w/cpp/language/range-for) and here is something with equivalent semantics: (I think are compiling in C++14 mode?)

```
for ( range_declaration : range_expression ) loop_statement
// is equivalent to
{
  auto && __range = range_expression ; 
  for (auto __begin = begin_expr, __end = end_expr; __begin != __end; ++__begin) { 
    range_declaration = *__begin; 
    loop_statement 
  } 
}
```

The move happens _before_ the iterator is created. So I would expected there not to be any iterator invalidation here.

Reply:
It's being moved during the re-allocation of `abiExportMap` as new entries get added to it

Reply:
Ah, I see what you mean! `addToABIExportMap` might create a new entry in the outer `DenseMap`, which moves the storage and hence moves the `DenseSet` in the value table.

Reply:
Does that sound correct? I should write a comment here explaining, otherwise one (possible future me) might be tempted to get rid of the copy.

Reply:
Yup, sounds good!

Reply:
I've refactored the code so that you don't call `addToABIExportMap` directly, instead the looping and iteration is handled for you. That also avoids the unnecessary copy by making sure to not having any moves in `addToABIExportMap`.